src/triage/component/postmodeling/contrast/group_evaluator.py

```diff
@@ -0,0 +1,325 @@
+routines that can be scaled up and grow according to the needs of the
+project, or other postmodeling approaches.
+
+To run most of this routines you will need:
+    - S3 credentials (if used) or specify the path to both feature and
+    prediction matrices.
+    - Working database db_engine.
+"""
+
+import pandas as pd
+import numpy as np
+from sqlalchemy.sql import text
+from matplotlib import pyplot as plt
+from collections import namedtuple
+from itertools import starmap, combinations
+from scipy.spatial.distance import squareform, pdist
+from scipy.stats import spearmanr
+import seaborn as sns
+
+from utils.file_helpers import download_s3
+from utils.test_conn import db_engine
+from utils.aux_funcs import recombine_categorical, pd_int_var
+
+# Get indivual model information/metadata from Audition output
+
+
+class ModelGroup(object):
+    '''
+    ModelGroup class calls the model group metadata from the database
+    and hold metadata features on each of the class attibutes.
+    This class will contain any information about the model_group, and 
+    will be used to make comparisons and calculations across models. 
+
+    A model_group_id list is needed to instate the class.
+    '''
+    def __init__(self, model_group_id):
+        self.model_group_id = model_group_id
+
+        # Retrive model_id metadata from the model_metadata schema
+        model_metadata = pd.read_sql("""
+        WITH
+        individual_model_ids_metadata AS(
+        SELECT m.model_id,
+               m.model_group_id,
+               m.hyperparameters,
+               m.model_hash,
+               m.experiment_hash,
+               m.train_end_time,
+               m.train_matrix_uuid,
+               m.training_label_timespan,
+               mg.model_type,
+               mg.model_config
+            FROM model_metadata.models m
+            JOIN model_metadata.model_groups mg
+            USING (model_group_id)
+            WHERE model_group_id IN {}
+        ),
+        individual_model_id_matrices AS(
+        SELECT DISTINCT ON (matrix_uuid)
+               model_id,
+               matrix_uuid
+            FROM test_results.predictions
+            WHERE model_id = ANY(
+                SELECT model_id
+                FROM individual_model_ids_metadata
+            )
+        )
+        SELECT metadata.model_id,
+               metadata.model_group_id,
+               metadata.model_hash,
+               metadata.hyperparameters,
+               metadata.train_end_time,
+               metadata.train_matrix_uuid,
+               metadata.training_label_timespan,
+               metadata.model_type,
+               metadata.model_config,
+               test.matrix_uuid AS test_matrix_uuid
+        FROM individual_model_ids_metadata AS metadata
+        LEFT JOIN individual_model_id_matrices AS test
+        USING(model_id);
+        """.format(self.model_group_id),
+                   con=db_engine).to_dict('list')
```

Comment:
I understand you want to populate **a bunch** of instance attributes; but, it's a little evil to eagerly read the database upon object construction, (before you actually need any of that info).

Otherwise, I would suggest, for example, using `descriptors.cachedproperty` from [Dickens](https://github.com/dssg/dickens):

    @cachedproperty
    def _model_metadata(self):
        return pd.read_sql("…")

    @cachedproperty
    def model_id(self):
        return self._model_metadata['model_id']

That said, this is a lot of attributes! So, I dunno.

BUT, I notice you're basically doing this, to an extent, with some other attributes, (which I'll get to shortly…).

Reply:
Agree